server/snapserver.1
.TH SNAPSERVER 1 "September 2025"
.SH NAME
snapserver - Snapcast server
.SH SYNOPSIS
\fBsnapserver\fR [\fIOPTION\fR]...
.SH DESCRIPTION
Snapcast is a multiroom client-server audio player, where all clients are
time synchronized with the server to play perfectly synced audio. It's not a
standalone player, but an extension that turns your existing audio player into
a Sonos-like multiroom solution.
.br
The server's audio input is a named pipe `/tmp/snapfifo`. All data that is fed
into this file will be send to the connected clients. One of the most generic
ways to use Snapcast is in conjunction with the music player daemon or Mopidy,
which can be configured to use a named pipe as audio output.
.SS Allowed options:
.TP
\fB-h, --help\fR
Produce help message, use -hh to show options from config file
.TP
\fB-v, --version\fR
Show version number
.TP
\fB-d, --daemon [=arg(=0)]\fR
Daemonize
optional process priority [-20..19]
.TP
\fB-c, --config arg (=/etc/snapserver.conf)\fR
Path to the configuration file
.SH FILES
.TP
\fI/tmp/snapfifo\fR
PCM input fifo file
.TP
\fI/etc/default/snapserver\fR
the daemon default configuration file
.TP
\fI/etc/snapserver.conf\fR
the snapserver configuration file
.TP
\fI~/.config/snapcast/server.json\fR or (if $HOME is not set) \fI/var/lib/snapcast/server.json\fR
persistent server data file
.SH "COPYRIGHT"
Copyright (C) 2014-2025 Johannes Pohl (snapcast@badaix.de).
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
snapclient(1)


changelog.md
# Snapcast changelog

## Version 0.33.0

### Features

- Server: Add PipeWire stream reader (PR #1418, Issue #1371)
- Client: Add PipeWire player (PR #1420, Issue #1419, #821)
- Server: process stream sources can be added via RPC (Issue #1359, PR #1444)
- Server: mDNS can be disabled (Issue #1172, PR #1437)
- Server: Audio streaming over plain TCP (port 1704) can be disabled (PR #1437)
- Server: Add control script for `go-librespot` (PR #1433, Issue #1424)

### Bugfixes

- Client: Fix case of command line arguments (Issue #1411)
- Client: Fix deadlock during shutdown (PR #1415)
- Fix building with boost v1.89.0 (Issues #1413)
- Fix MacOS CI (PR #1446)

### General

- Server: Log warning if default config is missing (PR #1402)
- Server: Make SSL support optional (`-DBUILD_WITH_SSL`) (Issue #1426, PR #1427)
- Make ALSA optional (PR #1422)
- Enable OpenSSL support for Android builds in `CMakeLists.txt` (PR #1373)
- Update Readme (PR #1431)
- Build Trixie packages for amd64 (Issue #1429)
- `tcp` section in `snapserver.conf` renamed to `tcp-control` (PR #1437)
- TCP streaming settings moved from `stream` to `tcp-streaming` (PR #1437)

### Contributors

- @badaix
- @mafrosis
- @aanno
- @filmor
- @sajjadsabzkar
- @MickMorley
- @lucianm

_Johannes Pohl <snapcast@badaix.de>  Tue, 23 Sep 2025 00:13:37 +0200_

## Version 0.32.3

### Features

### Bugfixes

- Server: re-add port to host in json RPC stream requests (Issue #1410)
- Documentation: Update Windows build instructions (PR #1408)

### General

### Contributors

- @stijnvdb88
- @badaix

_Johannes Pohl <snapcast@badaix.de>  Fri, 08 Aug 2025 00:13:37 +0200_

## Version 0.32.2

### Bugfixes

- Fix building Snapclient debian package with Pulse audio (Issue #1406)

### Contributors

- @badaix

_Johannes Pohl <snapcast@badaix.de>  Tue, 05 Aug 2025 00:13:37 +0200_

## Version 0.32.1

### Bugfixes

- Add OpenSSL dlls to Windows bundle (Issue #1405)

### Contributors

- @badaix

_Johannes Pohl <snapcast@badaix.de>  Mon, 04 Aug 2025 00:13:37 +0200_

## Version 0.32.0

### Features

- Client: Add support for (secure-) websockets (Issue #1325, PR #1340)
- Server: Add client certificate validation (Issue #1334)
- Server: Add control script for Plex (PR #1346)

### Bugfixes

- Fix typos (PR #1333, PR #1341, PR #1345)
- Add `mpv` option `--ao-pcm-waveheader=no` on player setup page (Issue #1094)
- Fix clang-tidy static analysis issues (Issue #1354)
- Fix building with Jack on Fedora (Issue #1363)
- Snapserver silently ignores if config file cannot be read (Issue #1402) 
- Add required packages to build instructions (Issue #1397)
- Fix stream URI string encoding (PR #1398)
- Fix reading port for TCP streams (PR #1399)

### General

- Snapweb: Update to v0.9.1
- Client: Command line arguments '--host' and '--port' are deprecated
- Update binary_protocol.md (Issue #1339)
- Add missing doxygen code documentation (Issue #1352)
- Add clang-tidy static analysis to CI (Issue #1353)
- Compiles with boost v1.88.0 (Issue #1367, PR #1378, PR #1387)
- Add Music Assistant project to readme (PR #1366)
 
### Contributors

- @chicco-carone
- @rnestler
- @badaix
- @chriscn
- @SantiagoSotoC
- @tobster
- @gnumpi

_Johannes Pohl <snapcast@badaix.de>  Sun, 03 Aug 2025 00:13:37 +0200_

## Version 0.31.0

### Features

- Server: Add secure RPC method `Stream.AddStream` (PR #1331, Issue #1330)
- Server: Re-add RPC method `RemoveStream` (Issue #1330)

### Bugfixes

- Fix crash when calling `Stream.RemoveStream`
- Fix typo in `changelog.md` (PR #1328)
- Fix typo in `configuration.md` (PR #1329)
- Update to AixLog v1.5.1, flushing log lines (PR #1332)

### General

- Add more doxygen documentation

_Johannes Pohl <snapcast@badaix.de>  Thu, 23 Jan 2025 00:13:37 +0200_

## Version 0.30.0

### Features

- Server: Add support for HTTPS and WSS (Issue #1322)
- Server: custom URL prefix for HTTP URLs (PR #1260)
- Server: listen to both IPv4 and IPv6 for incoming connections (PR #1290)

### Bugfixes

- Fix CVE-2023-36177: Remove "AddStream" and "RemoveStream" RPC (Issue #1327)
- Fix compile error on MacOS
- Remove PiMusicBox from readme (Issue ##1304)
- Add missing include (Issue #1295)
- Fix building with boost v1.87.0 (Issues #1305, #1306, #1307, PR #1308)
- Log tag is missing in syslog (Issue #1276)

### General

- CI: execute unit tests (asan and tsan)
- CI: create doxygen documentation
- CI: compile release and debug builds
- Code cleanup
- Add more doxygen documentation

_Johannes Pohl <snapcast@badaix.de>  Fri, 17 Jan 2025 00:13:37 +0200_

## Version 0.29.0

### Features

- Server: Add support for Jack streams (PR #1241, Issue #644)
- Client: Add support for 24 and 32 bit audio on Android

### Bugfixes

- Fix linker paths on ARM based MacOS

### General

- Snapweb: Update to v0.8.0
- CI: Add clang 16, 17, 18, gcc 14, xcode 15.1, 15.2, 15.3, 15.4
- CI: Create GitHub draft release

_Johannes Pohl <snapcast@badaix.de>  Sun, 25 Aug 2024 00:13:37 +0200_

## Version 0.28.0

### Features

- Server: Use filename as title for FileStreams
- Client: Loop through resolved server IPs during connect (Issue #1216)

### Bugfixes

- Server: Fix noise caused by reading half samples (Issue #1170)
- Server: Check open syscall error in PipeStream::do_connect (PR #1150)
- Server: Tweak Airplay support (#1102)
- Server: Lower log severity for shairport-sync (Issue #891)
- Server: Emits notifications for streams with codec=null (Is11sue #1205)
- Server: Fix delayed switching of meta streams (Issue #1187)
- Server: Mixer script calls piling up (Issue #1174)
- Server: Mixer script creates zombie processes (Issue #1157)
- Improve connect handling in meta_mopidy.py
- Improve connect handling in meta_librespot-java.py
- Fix cppcheck issues

### General

- CI: Build debian bookworm and bullseye packages for amd64, arm and arm64
- CI: Add cppcheck static analysis
- Update documentation (PR #1130, PR #1067)
- Delete deprecated Makefiles, CMake is the only supported buildsystem
- Snapweb: Update to v0.7.0


server/snapServer.cpp
/***
    This file is part of snapcast
    Copyright (C) 2014-2019  Johannes Pohl

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
***/

#include <chrono>
#include <memory>
#include <sys/resource.h>

#include "../externals/popl/include/popl.hpp"
#ifdef HAS_DAEMON
#include "common/daemon.h"
#endif
#include "common/sampleFormat.h"
#include "common/signalHandler.h"
#include "common/snapException.h"
#include "common/timeDefs.h"
#include "common/utils/string_utils.h"
#include "encoder/encoderFactory.h"
#include "message/message.h"
#include "server_settings.hpp"
#include "streamServer.h"
#if defined(HAS_AVAHI) || defined(HAS_BONJOUR)
#include "publishZeroConf/publishmDNS.h"
#endif
#include "aixlog.hpp"
#include "config.h"


volatile sig_atomic_t g_terminated = false;
std::condition_variable terminateSignaled;

using namespace std;
using namespace popl;


int main(int argc, char* argv[])
{
#ifdef MACOS
#pragma message "Warning: the macOS support is experimental and might not be maintained"
#endif

    int exitcode = EXIT_SUCCESS;
    try
    {
        ServerSettings settings;
        std::string pcmStream = "pipe:///tmp/snapfifo?name=default";
        std::string config_file = "/etc/snapserver.conf";

        OptionParser op("Allowed options");
        auto helpSwitch = op.add<Switch>("h", "help", "Produce help message");
        auto groffSwitch = op.add<Switch, Attribute::hidden>("", "groff", "produce groff message");
        auto versionSwitch = op.add<Switch>("v", "version", "Show version number");
#ifdef HAS_DAEMON
        int processPriority(0);
        auto daemonOption = op.add<Implicit<int>>("d", "daemon", "Daemonize\noptional process priority [-20..19]", 0, &processPriority);
        auto userValue = op.add<Value<string>>("", "user", "the user[:group] to run snapserver as when daemonized", "");
#endif

        op.add<Value<string>>("c", "config", "path to the configuration file", config_file, &config_file);

        // debug settings
        OptionParser conf("");
        conf.add<Switch>("", "logging.debug", "enable debug logging", &settings.logging.debug);
        conf.add<Value<string>>("", "logging.debug_logfile", "log file name for the debug logs (debug must be enabled)", settings.logging.debug_logfile,
                                &settings.logging.debug_logfile);

        // stream settings
        conf.add<Value<size_t>>("p", "stream.port", "Server port", settings.stream.port, &settings.stream.port);
        auto streamValue = conf.add<Value<string>>(
            "s", "stream.stream", "URI of the PCM input stream.\nFormat: TYPE://host/path?name=NAME\n[&codec=CODEC]\n[&sampleformat=SAMPLEFORMAT]", pcmStream,
            &pcmStream);

        conf.add<Value<string>>("", "stream.sampleformat", "Default sample format", settings.stream.sampleFormat, &settings.stream.sampleFormat);
        conf.add<Value<string>>("c", "stream.codec", "Default transport codec\n(flac|ogg|pcm)[:options]\nType codec:? to get codec specific options",
                                settings.stream.codec, &settings.stream.codec);
        conf.add<Value<size_t>>("", "stream.stream_buffer", "Default stream read buffer [ms]", settings.stream.streamReadMs, &settings.stream.streamReadMs);
        conf.add<Value<int>>("b", "stream.buffer", "Buffer [ms]", settings.stream.bufferMs, &settings.stream.bufferMs);
        conf.add<Switch>("", "stream.send_to_muted", "Send audio to muted clients", &settings.stream.sendAudioToMutedClients);

        // HTTP RPC settings
        conf.add<Switch>("", "http.enabled", "enable HTTP Json RPC (HTTP POST and websockets)", &settings.http.enabled);
        conf.add<Value<size_t>>("", "http.port", "which port the server should listen to", settings.http.port, &settings.http.port);
        conf.add<Value<string>>("", "http.doc_root", "serve a website from the doc_root location", settings.http.doc_root, &settings.http.doc_root);

        // TCP RPC settings
        conf.add<Switch>("", "tcp.enabled", "enable TCP Json RPC)", &settings.tcp.enabled);
        conf.add<Value<size_t>>("", "tcp.port", "which port the server should listen to", settings.tcp.port, &settings.tcp.port);

        // TODO: Should be possible to override settings on command line

        try
        {
            op.parse(argc, argv);
            conf.parse(config_file);
        }
        catch (const std::invalid_argument& e)
        {
            SLOG(ERROR) << "Exception: " << e.what() << std::endl;
            cout << "\n" << op << "\n";
            exit(EXIT_FAILURE);
        }

        if (versionSwitch->is_set())
        {
            cout << "snapserver v" << VERSION << "\n"
                 << "Copyright (C) 2014-2019 BadAix (snapcast@badaix.de).\n"
                 << "License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.\n"
                 << "This is free software: you are free to change and redistribute it.\n"
                 << "There is NO WARRANTY, to the extent permitted by law.\n\n"
                 << "Written by Johannes Pohl.\n";
            exit(EXIT_SUCCESS);
        }

        if (helpSwitch->is_set())
        {
            cout << op << "\n";
            exit(EXIT_SUCCESS);
        }

        if (groffSwitch->is_set())
        {
            GroffOptionPrinter option_printer(&op);
            cout << option_printer.print();
            exit(EXIT_SUCCESS);
        }

        if (settings.stream.codec.find(":?") != string::npos)
        {
            EncoderFactory encoderFactory;
            std::unique_ptr<Encoder> encoder(encoderFactory.createEncoder(settings.stream.codec));
            if (encoder)
            {
                cout << "Options for codec \"" << encoder->name() << "\":\n"
                     << "  " << encoder->getAvailableOptions() << "\n"
                     << "  Default: \"" << encoder->getDefaultOptions() << "\"\n";
            }
            exit(EXIT_SUCCESS);
        }

        AixLog::Log::init<AixLog::SinkNative>("snapserver", AixLog::Severity::trace, AixLog::Type::special);
        if (settings.logging.debug)
        {
            AixLog::Log::instance().add_logsink<AixLog::SinkCout>(AixLog::Severity::trace, AixLog::Type::all, "%Y-%m-%d %H-%M-%S.#ms [#severity] (#tag_func)");
            if (!settings.logging.debug_logfile.empty())
                AixLog::Log::instance().add_logsink<AixLog::SinkFile>(AixLog::Severity::trace, AixLog::Type::all, settings.logging.debug_logfile,
                                                                      "%Y-%m-%d %H-%M-%S.#ms [#severity] (#tag_func)");
        }
        else
        {
            AixLog::Log::instance().add_logsink<AixLog::SinkCout>(AixLog::Severity::info, AixLog::Type::all, "%Y-%m-%d %H-%M-%S [#severity]");
        }

        if (!streamValue->is_set())
            settings.stream.pcmStreams.push_back(streamValue->value());

        for (size_t n = 0; n < streamValue->count(); ++n)
        {
            LOG(INFO) << "Adding stream: " << streamValue->value(n) << "\n";
            settings.stream.pcmStreams.push_back(streamValue->value(n));
        }

        signal(SIGHUP, signal_handler);
        signal(SIGTERM, signal_handler);
        signal(SIGINT, signal_handler);

#ifdef HAS_DAEMON
        std::unique_ptr<Daemon> daemon;
        if (daemonOption->is_set())
        {
            string user = "";
            string group = "";

            if (userValue->is_set())
            {
                if (userValue->value().empty())
                    std::invalid_argument("user must not be empty");

                vector<string> user_group = utils::string::split(userValue->value(), ':');
                user = user_group[0];
                if (user_group.size() > 1)
                    group = user_group[1];
            }

            Config::instance().init("/var/lib/snapserver", user, group);
            daemon.reset(new Daemon(user, group, "/var/run/snapserver/pid"));
            daemon->daemonize();
            if (processPriority < -20)
                processPriority = -20;
            else if (processPriority > 19)
                processPriority = 19;
            if (processPriority != 0)
                setpriority(PRIO_PROCESS, 0, processPriority);
            SLOG(NOTICE) << "daemon started" << std::endl;
        }
        else
            Config::instance().init();
#else
        Config::instance().init();
#endif


#if defined(HAS_AVAHI) || defined(HAS_BONJOUR)
        PublishZeroConf publishZeroConfg("Snapcast");
        vector<mDNSService> dns_services;
        dns_services.emplace_back("_snapcast._tcp", settings.stream.port);
        dns_services.emplace_back("_snapcast-stream._tcp", settings.stream.port);
        if (settings.tcp.enabled)
        {
            dns_services.emplace_back("_snapcast-jsonrpc._tcp", settings.tcp.port);
            dns_services.emplace_back("_snapcast-tcp._tcp", settings.tcp.port);
        }
        if (settings.http.enabled)
        {
            dns_services.emplace_back("_snapcast-http._tcp", settings.http.port);
        }
        publishZeroConfg.publish(dns_services);
#endif

        if (settings.stream.bufferMs < 400)
        {
            LOG(WARNING) << "Buffer is less than 400ms, changing to 400ms\n";
            settings.stream.bufferMs = 400;
        }

        boost::asio::io_context io_context;
        std::unique_ptr<StreamServer> streamServer(new StreamServer(&io_context, settings));
        streamServer->start();

        auto func = [](boost::asio::io_context* ioservice) -> void { ioservice->run(); };
        std::thread t(func, &io_context);

        while (!g_terminated)
            chronos::sleep(100);

        io_context.stop();
        t.join();

        LOG(INFO) << "Stopping streamServer" << endl;
        streamServer->stop();
        LOG(INFO) << "done" << endl;
    }
    catch (const std::exception& e)
    {
        SLOG(ERROR) << "Exception: " << e.what() << std::endl;
        exitcode = EXIT_FAILURE;
    }

    SLOG(NOTICE) << "daemon terminated." << endl;
    exit(exitcode);
}
